File: services/baidu/tts/baidu_tts_client.cpp
#include "baidu/tts/baidu_tts_client.h"

#include <cstdio>
#include <stdexcept>

#include <httplib.h>

#include "async_logger.h"

namespace shizuru::services {

namespace {

// Percent-encode a UTF-8 string for use in application/x-www-form-urlencoded.
std::string UrlEncode(const std::string& s) {
  std::string result;
  result.reserve(s.size() * 3);  // worst case
  for (unsigned char c : s) {
    if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') ||
        (c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' || c == '~') {
      result.push_back(static_cast<char>(c));
    } else {
      char buf[4];
      std::snprintf(buf, sizeof(buf), "%%%02X", c);
      result.append(buf);
    }
  }
  return result;
}

}  // namespace

BaiduTtsClient::BaiduTtsClient(BaiduConfig config,
                               std::shared_ptr<BaiduTokenManager> token_mgr)
    : config_(std::move(config)), token_mgr_(std::move(token_mgr)) {}

std::string BaiduTtsClient::Synthesize(const std::string& text,
                                       std::string& mime_type) {
  LOG_INFO("[{}] Synthesize text_len={}", MODULE_NAME, text.size());

  std::string token = token_mgr_->GetToken();

  httplib::Client cli(config_.tts_host);
  cli.set_connection_timeout(config_.connect_timeout);
  cli.set_read_timeout(config_.read_timeout);

  // Build POST body (URL-encoded form).
  std::string body;
  body += "tex=" + UrlEncode(text);
  body += "&tok=" + token;
  body += "&cuid=" + config_.cuid;
  body += "&ctp=1";
  body += "&lan=zh";
  body += "&spd=" + std::to_string(config_.spd);
  body += "&pit=" + std::to_string(config_.pit);
  body += "&vol=" + std::to_string(config_.vol);
  body += "&per=" + std::to_string(config_.per);
  body += "&aue=" + std::to_string(config_.aue);

  auto res = cli.Post(config_.tts_path, body,
                      "application/x-www-form-urlencoded");

  if (!res) {
    LOG_ERROR("[{}] TTS request failed: no response", MODULE_NAME);
    throw std::runtime_error("Baidu TTS request failed: " +
                             httplib::to_string(res.error()));
  }

  if (res->status != 200) {
    LOG_ERROR("[{}] TTS status {}: {}", MODULE_NAME, res->status, res->body);
    throw std::runtime_error("Baidu TTS API returned status " +
                             std::to_string(res->status));
  }

  // Check Content-Type: audio/* means success; application/json means error.
  std::string content_type;
  if (res->has_header("Content-Type")) {
    content_type = res->get_header_value("Content-Type");
  }

  if (content_type.find("audio/") != std::string::npos) {
    // Determine MIME type from aue parameter.
    switch (config_.aue) {
      case 3:
        mime_type = "audio/mp3";
        break;
      case 4:
        mime_type = "audio/pcm";
        break;
      case 5:
        mime_type = "audio/pcm";
        break;
      case 6:
        mime_type = "audio/wav";
        break;
      default:
        mime_type = "audio/mp3";
        break;
    }
    LOG_INFO("[{}] TTS success, audio_len={} mime={}", MODULE_NAME,
             res->body.size(), mime_type);
    return res->body;
  }

  // Error response is JSON.
  LOG_ERROR("[{}] TTS error response: {}", MODULE_NAME, res->body);
  throw std::runtime_error("Baidu TTS returned error: " + res->body);
}

}  // namespace shizuru::services

File: ui/lib/bridge/baidu_tts_client.dart
// Baidu TTS client — text to audio (Dart port of C++ BaiduTtsClient).

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'baidu_token_manager.dart';

class BaiduTtsClient {
  final BaiduTokenManager tokenManager;
  final String ttsHost;
  final String ttsPath;
  final String cuid;
  final int spd;
  final int pit;
  final int vol;
  final int per;
  final int aue;

  BaiduTtsClient({
    required this.tokenManager,
    this.ttsHost = 'https://tsn.baidu.com',
    this.ttsPath = '/text2audio',
    this.cuid = 'shizuru_agent',
    this.spd = 5,
    this.pit = 5,
    this.vol = 5,
    this.per = 0,
    this.aue = 6, // 3=mp3, 4=pcm-8k, 5=pcm-16k, 6=wav
  });

  /// Synthesize [text] into audio bytes. Returns (audioBytes, mimeType).
  Future<(Uint8List, String)> synthesize(String text) async {
    final token = await tokenManager.getToken();

    final body = StringBuffer()
      ..write('tex=${Uri.encodeComponent(text)}')
      ..write('&tok=$token')
      ..write('&cuid=$cuid')
      ..write('&ctp=1')
      ..write('&lan=zh')
      ..write('&spd=$spd')
      ..write('&pit=$pit')
      ..write('&vol=$vol')
      ..write('&per=$per')
      ..write('&aue=$aue');

    final uri = Uri.parse('$ttsHost$ttsPath');
    final client = HttpClient()
      ..connectionTimeout = const Duration(seconds: 10);

    try {
      final request = await client.postUrl(uri);
      request.headers.contentType =
          ContentType('application', 'x-www-form-urlencoded');
      request.write(body.toString());
      final response = await request.close();

      if (response.statusCode != 200) {
        final errBody = await response.transform(utf8.decoder).join();
        throw Exception('Baidu TTS status ${response.statusCode}: $errBody');
      }

      final contentType =
          response.headers.value(HttpHeaders.contentTypeHeader) ?? '';

      final bytes = await _collectBytes(response);

      if (contentType.contains('audio/')) {
        final mime = _mimeFromAue(aue);
        return (bytes, mime);
      }

      // Error: response is JSON, not audio.
      final errText = utf8.decode(bytes);
      throw Exception('Baidu TTS error: $errText');
    } finally {
      client.close(force: true);
    }
  }

  String _mimeFromAue(int aue) {
    switch (aue) {
      case 3:
        return 'audio/mp3';
      case 4:
      case 5:
        return 'audio/pcm';
      case 6:
        return 'audio/wav';
      default:
        return 'audio/mp3';
    }
  }

  Future<Uint8List> _collectBytes(HttpClientResponse response) async {
    final builder = BytesBuilder(copy: false);
    await for (final chunk in response) {
      builder.add(chunk);
    }
    return builder.takeBytes();
  }
}
